### PR TITLE
sql/plan: unify iterative casts

### PR DESCRIPTION
**Pull request by sploiselle**
@JLDLaughlin Had some free time tonight and wanted to tinker around with this to see if I could get the iterative casts into the more standard casts (I couldn't), but had this nominal refactor sitting on my machine so figured I'd throw it up. If you already started this, glad to cede to your work and sorry if I stepped on your toes!

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4851)

### CODE REVIEW

File: src/sql/src/plan/typeconv.rs [hunk 1]
```diff
@@ -687,13 +634,20 @@ where
...
                 unsupported!("nested map types");
             }
-            match plan_cast_to_map(ecx, ccx, expr, &from_typ, &cast_to) {
+            if let ScalarExpr::CallUnary {
+                func: UnaryFunc::CastStringToMap { .. },
+                ..
+            } = expr
+            {
+                return Ok(expr);
+            };
```
Comment (benesch) [inline src/sql/src/plan/typeconv.rs:643]: This special case seems super weird to me!

**Review by benesch — APPROVED** [review]
This is great! Modulo one special case I find confusing.

File: src/sql/src/plan/typeconv.rs (hunk 1 above)
Comment (JLDLaughlin) [reply src/sql/src/plan/typeconv.rs:643]: This was me, not Sean!

I was going to write why I needed it, but I think we actually might not anymore. @sploiselle would you mind removing this test case and letting the tests run? I'm also happy to do this in a follow up PR!

File: src/sql/src/plan/typeconv.rs [hunk 2]
```diff
@@ -549,97 +554,39 @@ pub fn plan_cast_to_list(
...
+        (String, List(..)) => UnaryFunc::CastStringToList {
             return_ty,
             cast_expr,
         },
-        _ => unreachable!("already prevented match on incompatible types in plan_cast_to_list"),
-    }))
-}
-
-/// Plans a cast between from a [`ScalarType::String`] to [`ScalarType::Map`].
-pub fn plan_cast_to_map(
```
Comment (JLDLaughlin) [inline src/sql/src/plan/typeconv.rs:576]: 🙇

**Review by JLDLaughlin — APPROVED** [review]
This was first on my list today, thank you for reading my mind! This looks great!

File: src/sql/src/plan/typeconv.rs (hunk 1 above)
Comment (sploiselle) [reply src/sql/src/plan/typeconv.rs]: Should have rested its removal! Looks like it's superfluous now, so all good.

Comment (JLDLaughlin) [reply src/sql/src/plan/typeconv.rs]: Thank you for cleaning up after me!

**Pull request merged**